File: datastructures/CountingBloomFilter.java
package datastructures;
import java.util.*;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.nio.ByteBuffer;

public class CountingBloomFilter<T> {
  private final int[] counters;
  private final int size;
  private final int numHashFunctions;
  private final MessageDigest md;

  public CountingBloomFilter(int size, int numHashFunctions) {
    this.size = size;
    this.numHashFunctions = numHashFunctions;
    this.counters = new int[size];
    try {
      this.md = MessageDigest.getInstance("MD5");
    } catch (NoSuchAlgorithmException e) {
      throw new RuntimeException("MD5 not supported", e);
    }
  }

  public void add(T item) {
    for (int i = 0; i < numHashFunctions; i++) {
      int index = getHash(item, i);
      if (counters[index] < Integer.MAX_VALUE) {
        counters[index]++;
      }
    }
  }

  public void remove(T item) {
    for (int i = 0; i < numHashFunctions; i++) {
      int index = getHash(item, i);
      if (counters[index] > 0) {
        counters[index]--;
      }
    }
  }

  public boolean mightContain(T item) {
    for (int i = 0; i < numHashFunctions; i++) {
      if (counters[getHash(item, i)] == 0) {
        return false;
      }
    }
    return true;
  }

  private int getHash(T item, int i) {
    md.reset();
    md.update(item.toString().getBytes());
    md.update(ByteBuffer.allocate(4).putInt(i).array());
    byte[] digest = md.digest();
    return Math.abs(ByteBuffer.wrap(digest).getInt()) % size;
  }

  public double getEstimatedFalsePositiveRate(int numItems) {
    return Math.pow(1 - Math.exp(-numHashFunctions * (double) numItems / size), numHashFunctions);
  }

  // Stress test
  public static void main(String[] args) {
    int size = 1_000_000;
    int numHashFunctions = 5;
    CountingBloomFilter<String> filter = new CountingBloomFilter<>(size, numHashFunctions);

    int numItems = 100_000;
    Set<String> addedItems = new HashSet<>();

    long startTime = System.nanoTime();
    for (int i = 0; i < numItems; i++) {
      String item = UUID.randomUUID().toString();
      filter.add(item);
      addedItems.add(item);
    }
    long endTime = System.nanoTime();
    System.out.printf("Time to add %d items: %.2f ms%n", numItems, (endTime - startTime) / 1e6);

    int numTests = 1_000_000;
    int falsePositives = 0;
    startTime = System.nanoTime();
    for (int i = 0; i < numTests; i++) {
      String testItem = UUID.randomUUID().toString();
      if (filter.mightContain(testItem) && !addedItems.contains(testItem)) {
        falsePositives++;
      }
    }
    endTime = System.nanoTime();
    System.out.printf("Time to perform %d lookups: %.2f ms%n", numTests, (endTime - startTime) / 1e6);

    double actualFPRate = (double) falsePositives / numTests;
    double estimatedFPRate = filter.getEstimatedFalsePositiveRate(numItems);
    System.out.printf("False positive rate - Actual: %.6f, Estimated: %.6f%n", actualFPRate, estimatedFPRate);

    int numRemovals = numItems / 2;
    startTime = System.nanoTime();
    Iterator<String> iterator = addedItems.iterator();
    for (int i = 0; i < numRemovals; i++) {
      String item = iterator.next();
      filter.remove(item);
      iterator.remove();
    }
    endTime = System.nanoTime();
    System.out.printf("Time to remove %d items: %.2f ms%n", numRemovals, (endTime - startTime) / 1e6);

    int falseNegatives = 0;
    for (String item : addedItems) {
      if (!filter.mightContain(item)) {
        falseNegatives++;
      }
    }
    System.out.printf("False negatives after removal: %d (%.6f%%)%n",
        falseNegatives, (double) falseNegatives / addedItems.size() * 100);
  }
}

File: rust/src/bin/CountingBloomFilter.rs
use std::collections::HashSet;
use std::hash::Hash;
use std::time::Instant;
use uuid::Uuid;

pub struct CountingBloomFilter {
    counters: Vec<u32>,
    size: usize,
    num_hash_functions: usize,
}

impl CountingBloomFilter {
    pub fn new(size: usize, num_hash_functions: usize) -> Self {
        CountingBloomFilter {
            counters: vec![0; size],
            size,
            num_hash_functions,
        }
    }

    pub fn add<T: Hash>(&mut self, item: &T) {
        for i in 0..self.num_hash_functions {
            let index = self.get_hash(item, i);
            if self.counters[index] < u32::MAX {
                self.counters[index] += 1;
            }
        }
    }

    pub fn remove<T: Hash>(&mut self, item: &T) {
        for i in 0..self.num_hash_functions {
            let index = self.get_hash(item, i);
            if self.counters[index] > 0 {
                self.counters[index] -= 1;
            }
        }
    }

    pub fn might_contain<T: Hash>(&self, item: &T) -> bool {
        (0..self.num_hash_functions).all(|i| self.counters[self.get_hash(item, i)] > 0)
    }

    fn get_hash<T: Hash>(&self, item: &T, i: usize) -> usize {
        use std::hash::{Hash, Hasher};
        use std::collections::hash_map::DefaultHasher;

        let mut hasher = DefaultHasher::new();
        item.hash(&mut hasher);
        i.hash(&mut hasher);
        let hash = hasher.finish();

        let bytes = hash.to_le_bytes();
        let digest = md5::compute(bytes);
        let result = u64::from_le_bytes(digest[..8].try_into().unwrap());
        (result as usize) % self.size
    }

    pub fn get_estimated_false_positive_rate(&self, num_items: usize) -> f64 {
        (1.0 - (-((self.num_hash_functions as f64) * (num_items as f64) / (self.size as f64))).exp())
            .powi(self.num_hash_functions as i32)
    }
}

fn main() {
    let size = 1_000_000;
    let num_hash_functions = 5;
    let mut filter = CountingBloomFilter::new(size, num_hash_functions);

    let num_items = 100_000;
    let mut added_items = HashSet::new();

    let start_time = Instant::now();
    for _ in 0..num_items {
        let item = Uuid::new_v4();
        filter.add(&item);
        added_items.insert(item);
    }
    let end_time = Instant::now();
    println!(
        "Time to add {} items: {:.2} ms",
        num_items,
        end_time.duration_since(start_time).as_nanos() as f64 / 1e6
    );

    let num_tests = 1_000_000;
    let mut false_positives = 0;

    let start_time = Instant::now();
    for _ in 0..num_tests {
        let test_item = Uuid::new_v4();
        if filter.might_contain(&test_item) && !added_items.contains(&test_item) {
            false_positives += 1;
        }
    }
    let end_time = Instant::now();
    println!(
        "Time to perform {} lookups: {:.2} ms",
        num_tests,
        end_time.duration_since(start_time).as_nanos() as f64 / 1e6
    );

    let actual_fp_rate = false_positives as f64 / num_tests as f64;
    let estimated_fp_rate = filter.get_estimated_false_positive_rate(num_items);
    println!(
        "False positive rate - Actual: {:.6}, Estimated: {:.6}",
        actual_fp_rate, estimated_fp_rate
    );

    let num_removals = num_items / 2;
    let start_time = Instant::now();
    let mut iterator = added_items.iter();
    let mut removed_items = HashSet::new();
    for _ in 0..num_removals {
        if let Some(item) = iterator.next() {
            filter.remove(item);
            removed_items.insert(*item);
        }
    }
    let end_time = Instant::now();
    println!(
        "Time to remove {} items: {:.2} ms",
        num_removals,
        end_time.duration_since(start_time).as_nanos() as f64 / 1e6
    );

    let mut false_negatives = 0;
    for item in &removed_items {
        if filter.might_contain(item) {
            false_negatives += 1;
        }
    }
    println!(
        "False negatives after removal: {} ({:.6}%)",
        false_negatives,
        false_negatives as f64 / num_removals as f64 * 100.0
    );
}
